Beta refactor

PhilPJL opened the pull request:
More refactoring.
Mostly LifeTransmuter and SpellMaster.

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerPoolModifierBuilder.cs:
```diff
@@ -0,0 +1,75 @@
+            Definition.SetUsesDetermination(usesDetermination);
+            Definition.SetUsesAbilityScoreName(usesAbilityScoreName);
+            // This is just an activation time that should not be shown in the UI.
+            Definition.SetActivationTime(RuleDefinitions.ActivationTime.Permanent);
```
All of these hardcoded values should be set in the constructor. In particular making sure effect description is set and that cost per use is 1 (though really everything that's hardcoded should just be set in the constructor).

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerPoolModifierBuilder.cs:
```diff
@@ -0,0 +1,75 @@
+            Definition.SetEffectDescription(new EffectDescription());
+            Definition.SetOverriddenPower(Definition);
+
+            Definition.PoolPower = poolPower;
```
It should not be possible to construct a power pool modifier without providing a pool. I'd prefer either it is required in the constructor or we have validation to make sure it's been set.

PhilPJL replied:
Added validation

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerSharedPoolBuilder.cs:
```diff
@@ -13,6 +14,16 @@ public class FeatureDefinitionPowerPoolBuilder : FeatureDefinitionPowerBuilder<F
+        }
+
+        // TODO: move this to FeatureDefinitionPowerBuilder as ConfigurePowerPool()?
+        public FeatureDefinitionPowerPoolBuilder Configure(int usesPerRecharge,
```
Similar to other builder changes, I don't like that there is now a way to construct this without some of the values getting set (for example cost per use).

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerSharedPoolBuilder.cs:
```diff
@@ -64,6 +78,15 @@ public class FeatureDefinitionPowerSharedPoolBuilder : FeatureDefinitionPowerBui
+            Definition.SetGuiPresentation(guiPresentation);
+        }
+
+        public FeatureDefinitionPowerSharedPoolBuilder Configure(FeatureDefinitionPower poolPower,
```
I don't like that there is now a path to construct that allows some critical fields (uses determination, shared pool) to not be set.

ChrisPJohn commented on SolastaCommunityExpansion/Subclasses/Wizard/SpellMaster.cs:
```diff
@@ -101,61 +101,32 @@ internal SpellMaster()
-                        Category.Subclass, PowerWizardArcaneRecovery.GuiPresentation.SpriteReference));
-                BonusRecovery.SetCostPerUse(0);
-                BonusRecovery.SetRechargeRate(RuleDefinitions.RechargeRate.AtWill);
+                BonusRecovery
```
This might need a check that the BonusRecovery has been initialized.

PhilPJL replied:
I think all definitions have been created before anyone can view the settings page.

ThyWoof replied:
Settings won’t render until Main.Enabled equals true which happens at the very end of our init code.

ChrisPJohn commented on SolastaCommunityExpansion/Builders/ConditionDefinitionBuilder.cs:
```diff
@@ -207,10 +206,28 @@ public TBuilder SetDuration(RuleDefinitions.DurationType type, int duration = 0,
+            Definition.SetAllowMultipleInstances(false);
+            Definition.SetDurationType(durationType);
+            Definition.SetDurationParameter(durationParameter);
+            Definition.SetConditionStartParticleReference(new AssetReference());
```
Should these asset references be set in the initialise?

PhilPJL replied:
Yes, they were already being done in ctors.  Added same to Initialise.

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerBuilder.cs:
```diff
@@ -70,13 +70,30 @@ protected FeatureDefinitionPowerBuilder(TDefinition original, string name, strin
+
+            if (IsNew)
+            {
+                // This is just an activation time that should not be shown in the UI.
```
The activation time and costPerUse defaults here were only supposed to be for the power pool related builders that were previously setting them. I don't think they make sense in the base power builder.

PhilPJL replied:
moved back

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerPoolModifierBuilder.cs:
```diff
@@ -0,0 +1,70 @@
+            base.Validate();
+
+            Preconditions.IsNotNull(Definition.PoolPower, $"{GetType().Name}[{Definition.Name}].PoolPower is null.");
+            // TODO: more validation?
```
We should probably validate that costPerUse is 1 for the modifier.

PhilPJL replied:
done

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerSharedPoolBuilder.cs:
```diff
@@ -1,90 +1,89 @@
+
+            // We set uses determination to fixed because the code handling updates needs that.
+            Definition.SetUsesDetermination(RuleDefinitions.UsesDetermination.Fixed);
+            // Recharge rate probably shouldn't be in here, but for now leave it be because there is already usage outside of this mod.
```
Why is this comment here now?

PhilPJL replied:
cut and paste error

ChrisPJohn commented on SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerSharedPoolBuilder.cs:
```diff
@@ -1,90 +1,89 @@
-            Definition.SetEffectDescription(new EffectDescription());
-            Definition.SetOverriddenPower(Definition);
-            Definition.SetGuiPresentation(guiPresentation);
+            base.Validate();
```
This should validate that cost per use is 1 and uses determination is fixed.

PhilPJL replied:
done

ChrisPJohn replied:
So checking the old code outside of the git diff (because that's a nightmare). I was wrong about cost per use needing to be 1. However we really should validate that usesDetermination is set to fixed. I know it's in the initialize, but since the full power builder is available here it's quite possible a user could change that.

PhilPJL commented:
Closed to avoid further conflict.

PhilPJL commented:
I'll address issue raised previously

PhilPJL commented on SolastaCommunityExpansion/Builders/ConditionDefinitionBuilder.cs:
```diff
@@ -25,15 +24,38 @@ public abstract class ConditionDefinitionBuilder<TDefinition, TBuilder> : Defini
+        {
+            base.Initialise();
+
+            // NOTE: if a constructor is used Initialise won't get called from the DefinitionBuilder base
```
That was my original plan.  I've added a flag to disable the use of standard Create methods so the builder can force the use of more specialized Create method or constructor.

PhilPJL commented on SolastaCommunityExpansion/Builders/ConditionDefinitionBuilder.cs:
```diff
@@ -25,15 +24,38 @@ public abstract class ConditionDefinitionBuilder<TDefinition, TBuilder> : Defini
-                .SetConditionEndParticleReference(assetReference)
-                .SetCharacterShaderReference(assetReference);
+            // TODO: add initializer in extension generation
+            if (Definition.GetField<AssetReference>("conditionStartParticleReference") == null)
```
Added ClearParticleReferences and SetEmptyParticleReferencesWhereNull as extension methods.
Used SetEmptyParticleReferencesWhereNull in builder since we may copy an existing definition.

PhilPJL commented:
@ChrisPJohn 

added

```
        internal override void Validate()
        {
            base.Validate();

            Preconditions.IsNotNull(Definition.SharedPool, $"FeatureDefinitionPowerSharedPoolBuilder[{Definition.Name}].SharedPool is null.");
            **Preconditions.AreEqual(Definition.UsesDetermination, RuleDefinitions.UsesDetermination.Fixed, $"FeatureDefinitionPowerSharedPoolBuilder[{Definition.Name}].UsesDetermination must be set to Fixed.");**
        }
```

Does your comment above about CostPerUse mean I should remove 

`            Preconditions.AreEqual(Definition.CostPerUse, 1, $"{GetType().Name}[{Definition.Name}].CostPerUse must be set to 1.");
`

from `FeatureDefinitionPowerPoolModifierBuilder `and/or `FeatureDefinitionPowerPoolBuilder`?

ChrisPJohn commented:
> 

FeatureDefinitionPowerPoolModifierBuilder and FeatureDefinitionPowerPoolBuilder should both still verify cost per use is 1.

FeatureDefinitionPowerSharedPoolBuilder does not need to validate costPerUse (can be any).
